### ISSUE TITLE
Cannot set up "duration" field for "new experiment" of "Pod Kill"

### ISSUE THREAD

**Issue by GillesZhao** [chaos-mesh/chaos-mesh#1680]
![image](https://user-images.githubusercontent.com/58843253/114525831-02fc5800-9c79-11eb-8da6-9108d711149c.png)

**We cannot find where to fill in the "duration" field.
And if the "duration" equals to "immediate" or nothing filled in (api will get null/nil etc) , the pod chaos-controller-manager-xxxxxxxxx-xxxxx will fail with error crash.**

**And here're some error logs from the pod:**
```
[...]
2021-04-13T07:43:48.094Z        INFO    controller-runtime.controller   Starting Controller     {"controller": "stresschaos"}
2021-04-13T07:43:48.094Z        INFO    controller-runtime.controller   Starting workers        {"controller": "stresschaos", "worker count": 1}
2021-04-13T07:43:48.094Z        INFO    controller-runtime.controller   Starting Controller     {"controller": "stresschaos-scheduler-updater"}
2021-04-13T07:43:48.094Z        INFO    controller-runtime.controller   Starting workers        {"controller": "stresschaos-scheduler-updater", "worker count": 1}
**E0413 07:43:48.094376       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)**
goroutine 707 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x247df80, 0x3c3afa0)
        /tmp/go/pkg/mod/k8s.io/apimachinery@v0.17.1-beta.0/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /tmp/go/pkg/mod/k8s.io/apimachinery@v0.17.1-beta.0/pkg/util/runtime/runtime.go:48 +0x82
panic(0x247df80, 0x3c3afa0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/chaos-mesh/chaos-mesh/controllers/twophase.(*Reconciler).Reconcile(0xc000884500, 0xc0008c99e0, 0xa, 0xc0009a03e0, 0x18, 0xc0009f3300, 0x0, 0x0, 0x0)
        /mnt/controllers/twophase/types.go:93 +0x1111
github.com/chaos-mesh/chaos-mesh/pkg/router.(*Reconciler).Reconcile(0xc00047f3b0, 0xc0008c99e0, 0xa, 0xc0009a03e0, 0x18, 0x0, 0xc01570cd059587ab, 0xc0005b01b0, 0xc0005b0128)
        /mnt/pkg/router/reconciler.go:109 +0x883
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000f2600, 0x251f340, 0xc000679920, 0xc0004e8600)
        /tmp/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256 +0x161
[...]
```

**Comment by AsterNighT** [comment]
Hi, thank for reporting this. Can you please provide you chaosmesh version? And how did you create the `podkillchaos` with "duration" equals to "immediate" or nothing filled, By chaos dashboard or `kubectl apply -f`?

**Comment by GillesZhao** [comment]
We did it by chaos dashboard, you can see my screenshot for reference. After we filled in the cron expressions we cannot see the "duration" field, and then we click "submit", and we just see duration equals to "immediate" 👍 

![image](https://user-images.githubusercontent.com/58843253/114533131-2aa2ee80-9c80-11eb-96c8-fd3752833ace.png)

and then the pod chaos-controller-manager-xxxxxxxxx-xxxxx will fail with error crash.

**Comment by AsterNighT** [comment]
![image](https://user-images.githubusercontent.com/22937027/114533336-5e7e1400-9c80-11eb-956a-9528cfdd1477.png)
I checked chaos dashboard and seem to find it. Could you please check your chaosmesh version?

**Comment by GillesZhao** [comment]
We are using the latest version v1.1.2

**Comment by GillesZhao** [comment]
The issue occurs from time to time, sometimes we see the "duration" sometimes it disappeared.

**Comment by AsterNighT** [comment]
@g1eny0ung PTAL

**Comment by g1eny0ung** [comment]
Due to the `pod-kill` action will delete the pod immediately, there is no needed to fill the duration field when you create an experiment.

And once the action `pod-kill` or  `container-kill` be selected, the duration field will be hidden automatically. This is an expected behavior.

**Comment by g1eny0ung** [comment]
I think we will describe this behavior more obviously, both in the dashboard interface and docs.

**Comment by GillesZhao** [comment]
But why the "duration" for pod-kill and container-kill appears and disappears from time to time? And even if the duration field is hidden, and the value will equal to "immediate", then the pod chaos-controller-manager-xxxxxxxxx-xxxxx will fail with error crash, it does not accept the value of "immediate".

**Comment by g1eny0ung** [comment]
> But why the "duration" for pod-kill and container-kill appears and disappears from time to time(Please refer to your colleague @AsterNighT 's screenshot and the one of mine)? And even if the duration field is hidden, and the value will equal to "immediate", then the pod chaos-controller-manager-xxxxxxxxx-xxxxx will fail with error crash, it does not accept the value of "immediate".

Aster's screenshot represents a temporary state, he has not yet chosen the final action. Only after the action is selected, the duration field will judge whether to hide according to the selection.

The **Duration immediate** shown on the dashboard is only related to UI. The below is its code position:

https://github.com/chaos-mesh/chaos-mesh/blob/master/ui/src/components/ExperimentConfiguration/index.tsx#L141

It means that the **immediate** string is displayed when the duration is empty.

At last, why the dashboard occurs a nil panic error, I have to think for a while 🤔.

**Comment by g1eny0ung** [comment]
@GillesZhao Can you provide the output of `kubectl get StressChaos -n chaos-testing`?

**Comment by GillesZhao** [comment]
> @GillesZhao Can you provide the output of `kubectl get StressChaos -n chaos-testing`?

Could you describe what kind of logs do you want to see, since actually I have no resource name for StressChaos and the namespace chaos-testing?

**Comment by g1eny0ung** [comment]
My fault 🤣, the command I want you to execute is `kubectl get StressChaos -A`. I'm just want to see if there has some failed stresschaos. Because I found the issue #802 is similar to your problem.

**Comment by GillesZhao** [comment]
Actually we only tried experiments Pod and Network. We never created one for stresschaos.

**Comment by g1eny0ung** [comment]
🤔 I may not be able to figure out the problem. @cwen0 @AsterNighT Would u please take a look if you have free time?

**Comment by AsterNighT** [comment]
Yeah I am able to reproduce this. 
1. Select `pod kill` and confirm grace period. At this time there's no duration field
2. Cancel the configuration and select something else, like `pod failure`. Now duration should come up.
3. Cancel the configuration, select `pod kill` and confirm grace period again. Duration field won't disappear here.

@g1eny0ung I suppose that'll be some frontend bug?

**Comment by AsterNighT** [comment]
But it won't let me pass if I fill in some `immediate` or nothing in the duration field.

**Comment by g1eny0ung** [comment]
Since there is no feedback for a time, I will close this issue.

The above PR contains a tiny enhancement about the `immediate` note on the detail page. It will be removed when an experiment has no duration field.
a
If you still have some other problems, feel free to open a new issue.
